securesystemslib/hsm_keys.py

```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/env python
+
+"""
+<Program Name>
+  hsm_keys.py
+
+<Author>
+  Tanishq Jasoria <jasoriatanishq@gmail.com>
+
+<Purpose>
+  This module provides a high-level API for using hardware security modules
+  for various cryptographic operations
+
+  This module currently supports
+  1. Create and Verify signature using keys from a HSM
+  2. Export public key and X509 certificates stored in HSM in PEM format.
+"""
+
+from securesystemslib.hsm import HSM
+import securesystemslib.interface
+import securesystemslib.exceptions
+import binascii
+
+# Import cryptography routines needed to retrieve cryptographic
+# keys and certificates in PEM format.
+CRYPTO = True
+NO_CRYPTO_MSG = "To retrieve cryptographic keys and certificates in PEM format " \
+                "cryptography library is needed"
+try:
+  from cryptography.hazmat.primitives import serialization
+  from cryptography.hazmat.backends import default_backend
+except ImportError:
+  CRYPTO = False
+
+
+def load_library(PKCS11LIB_USER):
+  """
+  <Purpose>
+    To load a custom library to interact with the hardware tokens.
+
+  <Exceptions>
+    securesystemslib.exceptions.NotFoundError, if the path of PKCS#11
+    library is not specified or the library is corrupt.
+  """
+
+  global smartcard
+  smartcard = HSM(PKCS11LIB_USER)
+
+
+def load_HSMs():
+  """
+  <Purpose>
+    To get list of all the available HSMs
+
+  <Exceptions>
+    securesystemslib.exceptions.NotFoundError, if the path of PKCS#11
+    library is not specified in 'settings.py' or the library is corrupt
+
+    securesystemslib.exceptions.Error, if the function call is made
+    before the initialization call using load_library.
+
+  <Returns>
+    list of dictionaries corresponding to all the available HSMs
+  """
+
+  # All the functions must use the same object of the HSM class,
+  # to use same session for all the operations.
+
+  # Get information regarding the available HSM
+
+  try:
+    available_HSM = smartcard.get_available_HSMs()
+  except NameError:
+    raise securesystemslib.exceptions.Error(
+        'securesystemslib.hsm_key.load_library call must be used '
+        'to initialize the module, before reading the list of HSMs')
+
+  return available_HSM
+
+
+
+
+
+def load_private_keys(HSM_info, user_pin):
+  """
+  <Purpose>
+    Get list of handles of private keys stored in the HSM
+    corresponding to the 'HSM_info'
+
+  <Arguments>
+    HSM_info:
+      element from the list generated by 'load_HSMs()'
+    user_pin:
+      user pin of the HSM, to access private object from the token
+
+  <Exceptions>
+    securesystemlib.exceptions.InvalidNameError, if the requested
+    HSM is either not present or cannot be used.
+
+    securesystemslib.exceptions.BadPasswordError, if the entered
+    user pin is invalid.
+
+  <Returns>
+    list of handle of all the available private keys
+  """
+
+  # Use the HSM with the corresponding 'slot_info'
+  smartcard.get_HSM_session(HSM_info)
```

Comment:
If you call this function before `load_library` you will get `NameError: name 'smartcard' is not defined`

Reply:
Same comment applies to other functions that call `smartcard`

Reply:
Yes, there is a particular flow that needs to be followed while using this. 
It should give a more meaningful error to the user, I would change that.